src/WebJobs.Script.WebHost/DependencyInjection/DependencyValidator/ServiceMatch.cs

```diff
@@ -0,0 +1,181 @@
+            string fullTypeName = peerType.AssemblyQualifiedName.Replace(peerType.FullName, typeName);
+
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, Type.GetType(fullTypeName), lifetime);
+            _descriptors.Add(desc);
+        }
+
+        public void AddFactory<TFactoryType>(ServiceLifetime lifetime)
+        {
+            // Hijack the factory to pass in the expected type
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, s => typeof(TFactoryType).Assembly, lifetime);
+            _descriptors.Add(desc);
+        }
+
+        public bool Validate(IServiceCollection services)
+        {
+            IEnumerable<ServiceDescriptor> registered = services.Where(p => p.ServiceType == ServiceType);
+
+            switch (_match)
+            {
+                case MatchType.None:
+                    return !registered.Any();
+
+                case MatchType.Single:
+                    ServiceDescriptor expected = _descriptors.Single();
+                    ServiceDescriptor lastMatch = registered.LastOrDefault();
+                    return IsMatch(expected, lastMatch);
+
+                case MatchType.Collection:
+                    return IsCollection(registered, _descriptors);
+
+                case MatchType.Subcollection:
+                    return IsSubcollection(registered, _descriptors);
+
+                default:
+                    throw new InvalidOperationException($"Invalid value: '{_match}'");
+            }
+        }
+
+        private static bool IsMatch(ServiceDescriptor expected, ServiceDescriptor registered)
+        {
+            bool factoryMatches = false;
+
+            if (expected.ImplementationFactory != null && registered.ImplementationFactory != null)
+            {
+                // Make sure the factory is declared in the assembly mentioned; that's enough
+                // to make sure that it's not been overridden. Use the hijacked factory from
+                // when we registered this descriptor to pull out the expected Assembly.
+                Assembly expectedAssembly = (Assembly)expected.ImplementationFactory(null);
+                factoryMatches = expectedAssembly == registered.ImplementationFactory.GetMethodInfo().DeclaringType.Assembly;
+            }
+            else
+            {
+                factoryMatches = expected.ImplementationFactory == registered.ImplementationFactory;
+            }
+
+            return factoryMatches &&
+                expected.ImplementationInstance == registered.ImplementationInstance &&
+                expected.ImplementationType == registered.ImplementationType &&
+                expected.Lifetime == registered.Lifetime &&
+                expected.ServiceType == registered.ServiceType;
+        }
+
+        private static bool IsCollection(IEnumerable<ServiceDescriptor> registeredServices, IEnumerable<ServiceDescriptor> expectedServices)
```

Comment:
May want to take a list/collection or materialize these enumerables as they're used multiple times